### PR TITLE
lifecycle: add unit test for IBU.

### PR DESCRIPTION
**Pull request by achuzhoy**

### CODE REVIEW

File: pkg/lca/ibu_test.go [hunk 1]
```diff
@@ -0,0 +1,753 @@
...
+	assert.Equal(t, "", testBuilder.errorMsg)
+
+	testBuilder = testBuilder.WithOptions(
+		func(builder *ImageBasedUpgradeBuilder) (*ImageBasedUpgradeBuilder, error) {
+			return builder, fmt.Errorf("error")
+		})
+	assert.Equal(t, "error", testBuilder.errorMsg)
+}
+
+func buildTestClientWithDummyObject() *clients.Settings {
```
Comment (kononovn) [inline pkg/lca/ibu_test.go:41]: Please move all private function to the end of the file

File: pkg/lca/ibu_test.go [hunk 2]
```diff
@@ -0,0 +1,753 @@
...
+		},
+		Spec: lcav1alpha1.ImageBasedUpgradeSpec{
+			Stage: "Idle",
+		},
+	})
+}
+
+func TestImageBasedUpgradePull(t *testing.T) {
+	testCases := []struct {
+		expectedError       bool
```
Comment (kononovn) [inline pkg/lca/ibu_test.go:61]: Change it to expectedError err.
Then you can get rid of if/else

File: pkg/lca/ibu_test.go [hunk 3]
```diff
@@ -0,0 +1,753 @@
...
+				assert.Equal(t, testCase.expectedErrorText, err.Error())
+			}
+		} else {
+			assert.Nil(t, err)
+			assert.NotNil(t, builderResult)
+		}
+	}
+}
+
+func TestImageBasedUpgradeUpdate(t *testing.T) {
```
Comment (kononovn) [inline pkg/lca/ibu_test.go:104]: Please check reference below:

```
func TestArgoCdUpdate(t *testing.T) {
	testCases := []struct {
		testArgoCd    *Builder
		expectedError error
		image         string
	}{
		{
			testArgoCd:    buildValidArgoCdBuilder(buildArgoCdTestClientWithDummyObject()),
			expectedError: nil,
			image:         "testimage",
		},
		{
			testArgoCd:    buildInValidArgoCdBuilder(buildArgoCdTestClientWithDummyObject()),
			expectedError: fmt.Errorf("argocd 'nsname' cannot be empty"),
			image:         "testimage",
		},
		{
			testArgoCd:    buildValidArgoCdBuilder(clients.GetTestClients(clients.TestClientParams{})),
			expectedError: fmt.Errorf("argocds.argoproj.io \"argocd\" not found"),
		},
	}

	for _, testCase := range testCases {
		assert.Equal(t, "", testCase.testArgoCd.Definition.Spec.Image)
		testCase.testArgoCd.Definition.ResourceVersion = "999"
		assert.Equal(t, "", testCase.testArgoCd.Definition.Spec.Image)
		testCase.testArgoCd.Definition.Spec.Image = testCase.image
		_, err := testCase.testArgoCd.Update(false)

		if errors.IsNotFound(err) {
			assert.Equal(t, testCase.expectedError.Error(), err.Error())
		} else {
			assert.Equal(t, testCase.expectedError, err)
		}

		if testCase.expectedError == nil {
			assert.Equal(t, testCase.image, testCase.testArgoCd.Object.Spec.Image)
			assert.Equal(t, testCase.testArgoCd.Object, testCase.testArgoCd.Definition)
		}
	}
}
```

**Review by kononovn — CHANGES_REQUESTED** [review]
Please check few comments

File: pkg/lca/ibu_test.go (hunk 2 above)
Comment (achuzhoy) [reply pkg/lca/ibu_test.go]: Done.

File: pkg/lca/ibu_test.go (hunk 1 above)
Comment (achuzhoy) [reply pkg/lca/ibu_test.go]: Done.

File: pkg/lca/ibu_test.go (hunk 3 above)
Comment (achuzhoy) [reply pkg/lca/ibu_test.go:81]: I don't see how its a good reference.
testArgoCd is a pointer to Builder in the struct.

to set testArgoCd we then call  buildValidArgoCdBuilder which calls the NewBuilder function.
lca doesn't have this function

File: pkg/lca/ibu_test.go [hunk 4]
```diff
@@ -0,0 +1,693 @@
...
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "upgrade",
+		},
+	}
+}
+
+func buildTestClientWithDummyObject() *clients.Settings {
+	return clients.GetTestClients(clients.TestClientParams{
+		K8sMockObjects: buildDummyIBU(),
+		GVK:            []schema.GroupVersionKind{ibulGvk},
```
Comment (trewest) [inline pkg/lca/ibu_test.go:680]: Why is this needed?

File: pkg/lca/ibu_test.go [hunk 5]
```diff
@@ -0,0 +1,693 @@
...
+		{
+			expectedError:       nil,
+			addToRuntimeObjects: true,
+			stage:               "Idle",
+		},
+		{
+			expectedError:       nil,
+			addToRuntimeObjects: true,
+			stage:               "Wrong",
+		},
```
Comment (trewest) [inline pkg/lca/ibu_test.go:634]: Not sure if you really need this. It doesn't really validate anything different than the first test. Unless you are expecting to change it in the future

File: pkg/lca/ibu_test.go [hunk 6]
```diff
@@ -0,0 +1,693 @@
...
+			expectedError:       fmt.Errorf("wrong stage selected for imagebasedupgrade"),
+			addToRuntimeObjects: true,
+			stage:               "wrong",
+		},
+		{
+			expectedError:       fmt.Errorf("wrong stage selected for imagebasedupgrade"),
+			addToRuntimeObjects: true,
+			stage:               "",
+		},
+	}
```
Comment (trewest) [inline pkg/lca/ibu_test.go:584]: You should include a positive test no?

File: pkg/lca/ibu_test.go [hunk 7]
```diff
@@ -0,0 +1,693 @@
...
+			assert.NotNil(t, builderResult)
+		}
+	}
+}
+
+func TestImageBasedUpgradeWithSeedImagePullSecretRef(t *testing.T) {
+	testCases := []struct {
+		expectedError       error
+		addToRuntimeObjects bool
+		expectedErrorText   string
```
Comment (trewest) [inline pkg/lca/ibu_test.go:520]: Unused

File: pkg/lca/ibu_test.go [hunk 8]
```diff
@@ -0,0 +1,693 @@
...
+			assert.NotNil(t, builderResult)
+		}
+	}
+}
+
+func TestImageBasedUpgradeWithSeedImageVersion(t *testing.T) {
+	testCases := []struct {
+		expectedError       error
+		addToRuntimeObjects bool
+		expectedErrorText   string
```
Comment (trewest) [inline pkg/lca/ibu_test.go:469]: Unused

File: pkg/lca/ibu_test.go [hunk 9]
```diff
@@ -0,0 +1,693 @@
...
+			assert.NotNil(t, builderResult)
+		}
+	}
+}
+
+func TestImageBasedUpgradeWithOadpContent(t *testing.T) {
+	testCases := []struct {
+		expectedError                 error
+		addToRuntimeObjects           bool
+		expectedErrorText             string
```
Comment (trewest) [inline pkg/lca/ibu_test.go:415]: Unused

File: pkg/lca/ibu_test.go [hunk 10]
```diff
@@ -0,0 +1,693 @@
...
+			assert.NotNil(t, builderResult)
+		}
+	}
+}
+
+func TestImageBasedUpgradeWithExtraManifests(t *testing.T) {
+	testCases := []struct {
+		expectedError                    error
+		addToRuntimeObjects              bool
+		expectedErrorText                string
```
Comment (trewest) [inline pkg/lca/ibu_test.go:361]: Unused

File: pkg/lca/ibu_test.go [hunk 11]
```diff
@@ -0,0 +1,693 @@
...
+			assert.NotNil(t, builderResult)
+		}
+	}
+}
+
+func TestImageBasedUpgradeWithAdditionalImages(t *testing.T) {
+	testCases := []struct {
+		expectedError                      error
+		addToRuntimeObjects                bool
+		expectedErrorText                  string
```
Comment (trewest) [inline pkg/lca/ibu_test.go:307]: Unused

File: pkg/lca/ibu_test.go [hunk 12]
```diff
@@ -0,0 +1,693 @@
...
+			assert.NotNil(t, builderResult)
+		}
+	}
+}
+
+func TestImageBasedUpgradeWithSeedImage(t *testing.T) {
+	testCases := []struct {
+		expectedError       error
+		addToRuntimeObjects bool
+		expectedErrorText   string
```
Comment (trewest) [inline pkg/lca/ibu_test.go:257]: Unused

File: pkg/lca/ibu_test.go [hunk 13]
```diff
@@ -0,0 +1,693 @@
...
+			assert.NotNil(t, builderResult)
+		}
+	}
+}
+
+func TestImageBasedUpgradeExists(t *testing.T) {
+	testCases := []struct {
+		expectedError       error
+		addToRuntimeObjects bool
+		expectedErrorText   string
```
Comment (trewest) [inline pkg/lca/ibu_test.go:214]: Unused

File: pkg/lca/ibu_test.go [hunk 14]
```diff
@@ -0,0 +1,693 @@
...
+	}{
+		{
+			expectedError:       nil,
+			addToRuntimeObjects: true,
+		},
```
Comment (trewest) [inline pkg/lca/ibu_test.go:219]: I think you should include a negative test

File: pkg/lca/ibu_test.go [hunk 15]
```diff
@@ -0,0 +1,693 @@
...
+func TestImageBasedUpgradeGet(t *testing.T) {
+	testCases := []struct {
+		expectedError       error
+		addToRuntimeObjects bool
+		expectedErrorText   string
+	}{
+		{
+			expectedError:       nil,
+			addToRuntimeObjects: true,
+		},
```
Comment (trewest) [inline pkg/lca/ibu_test.go:176]: I think you should include a negative test

File: pkg/lca/ibu_test.go [hunk 16]
```diff
@@ -0,0 +1,693 @@
...
+			assert.Nil(t, builderResult.Object)
+		}
+	}
+}
+
+func TestImageBasedUpgradeGet(t *testing.T) {
+	testCases := []struct {
+		expectedError       error
+		addToRuntimeObjects bool
+		expectedErrorText   string
```
Comment (trewest) [inline pkg/lca/ibu_test.go:171]: Unused

File: pkg/lca/ibu_test.go [hunk 17]
```diff
@@ -0,0 +1,693 @@
...
+			assert.NotNil(t, builderResult)
+		}
+	}
+}
+
+func TestImageBasedUpgradeDelete(t *testing.T) {
+	testCases := []struct {
+		expectedError       error
+		addToRuntimeObjects bool
+		expectedErrorText   string
```
Comment (trewest) [inline pkg/lca/ibu_test.go:128]: Unused

File: pkg/lca/ibu_test.go [hunk 18]
```diff
@@ -0,0 +1,693 @@
...
+
+		// Test the PullImageBasedUpgrade function
+		builderResult, err := ibuBuilder.WithSeedImage("quay.io/no-image").Update()
+
+		// Check the error
+		assert.Equal(t, err, testCase.expectedError)
+
+		if testCase.expectedError == nil {
+			assert.NotNil(t, builderResult)
+		}
```
Comment (trewest) [inline pkg/lca/ibu_test.go:120]: I think you should validate the value of spec.seedImageRef.image is what you set it to

Comment (achuzhoy) [reply pkg/lca/ibu_test.go:120]: Added

File: pkg/lca/ibu_test.go (hunk 17 above)
Comment (achuzhoy) [reply pkg/lca/ibu_test.go:128]: Removed.

File: pkg/lca/ibu_test.go (hunk 16 above)
Comment (achuzhoy) [reply pkg/lca/ibu_test.go:171]: Removed.

File: pkg/lca/ibu_test.go (hunk 13 above)
Comment (achuzhoy) [reply pkg/lca/ibu_test.go:214]: Removed.

File: pkg/lca/ibu_test.go (hunk 12 above)
Comment (achuzhoy) [reply pkg/lca/ibu_test.go:257]: Removed.

File: pkg/lca/ibu_test.go (hunk 11 above)
Comment (achuzhoy) [reply pkg/lca/ibu_test.go:307]: Removed.

File: pkg/lca/ibu_test.go (hunk 10 above)
Comment (achuzhoy) [reply pkg/lca/ibu_test.go:361]: Removed.

File: pkg/lca/ibu_test.go (hunk 9 above)
Comment (achuzhoy) [reply pkg/lca/ibu_test.go:415]: Removed.

File: pkg/lca/ibu_test.go (hunk 8 above)
Comment (achuzhoy) [reply pkg/lca/ibu_test.go:469]: Removed.

File: pkg/lca/ibu_test.go (hunk 6 above)
Comment (achuzhoy) [reply pkg/lca/ibu_test.go:576]: A positive test will cause a jump into the PollUntilContextTimeout which takes 30 minutes

File: pkg/lca/ibu_test.go (hunk 15 above)
Comment (achuzhoy) [reply pkg/lca/ibu_test.go:175]: Added

File: pkg/lca/ibu_test.go (hunk 14 above)
Comment (achuzhoy) [reply pkg/lca/ibu_test.go:223]: Added

File: pkg/lca/ibu_test.go (hunk 5 above)
Comment (achuzhoy) [reply pkg/lca/ibu_test.go:641]: changed.

File: pkg/lca/ibu_test.go (hunk 4 above)
Comment (achuzhoy) [reply pkg/lca/ibu_test.go:687]: It's not needed - removed

File: pkg/lca/ibu_test.go (hunk 6 above)
Comment (trewest) [reply pkg/lca/ibu_test.go:580]: It appears to me that it would take just a second if you use `buildTestClientWithDummyObject`

Comment (achuzhoy) [reply pkg/lca/ibu_test.go:579]: Updated. Thanks.

**Review by trewest — APPROVED** [review]
/lgtm
